src/Npgsql/NpgsqlSlimDataSourceBuilder.cs

```diff
@@ -239,6 +242,34 @@ public NpgsqlSlimDataSourceBuilder UseRootCertificateCallback(Func<X509Certifica
         return this;
     }
 
+    /// <summary>
+    /// Configures a password provider, which is called by the data source when opening connections.
+    /// </summary>
+    /// <param name="passwordProvider">
+    /// A callback that may be invoked during <see cref="NpgsqlConnection.Open()" /> which returns the password to be sent to PostgreSQL.
+    /// </param>
+    /// <param name="passwordProviderAsync">
+    /// A callback that may be invoked during <see cref="NpgsqlConnection.OpenAsync(CancellationToken)" /> which returns the password to be sent to PostgreSQL.
+    /// </param>
+    /// <returns>The same builder instance so that multiple calls can be chained.</returns>
+    /// <remarks>
+    /// <para>
+    /// The provided callback is invoked when opening connections. Therefore its important the callback internally depends on cached
+    /// data or returns quickly otherwise. Any unnecessary delay will affect connection opening time.
+    /// </para>
+    /// </remarks>
+    public NpgsqlSlimDataSourceBuilder UsePasswordProvider(
+        Func<NpgsqlConnectionStringBuilder, string>? passwordProvider,
+        Func<NpgsqlConnectionStringBuilder, CancellationToken, ValueTask<string>>? passwordProviderAsync)
+    {
+        if (passwordProvider is null != passwordProviderAsync is null)
```

Comment:
OK, so either you provide none or you provide both.

I remember we discussed this already; another option here would be to allow you to not provide e.g. sync, and to throw if Open() is invoked - that would be nicer for the modern async-only crowd, which wouldn't need to provide a sync callback. On the other hand they can just provide a lambda which throws, which is very similar to passing null.

So am OK either way.

Reply:
As discussed offline we can always loosen it later.